staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,2147 @@
+
+        // If we receive an error message check if that's a busoff.
+        // If no just drop the message, otherwise update config and return.
+        if (response.can_id.to_native() & CAN_ERR_FLAG) != 0 {
+            if (response.can_id.to_native() & CAN_ERR_BUSOFF) != 0 {
+                // TODO: Trigger a config_change notification
+                self.controller.write().unwrap().config.status = VIRTIO_CAN_S_CTRL_BUSOFF.into();
+                self.controller.write().unwrap().ctrl_state = CAN_CS_STOPPED;
+                warn!("Got BusOff error frame, device does a local bus off\n");
+                return Err(Error::BusoffRxFrame);
+            } else {
+                trace!("Dropping error frame 0x{:x}\n", response.can_id.to_native());
+                return Err(Error::RxFrameUnknownFail);
+            }
+        }
+
+        // One of VIRTIO_CAN_F_CAN_CLASSIC and VIRTIO_CAN_F_CAN_FD must be negotiated
+        if (response.flags.to_native() & CAN_FRMF_TYPE_FD) != 0 {
+            if !self.check_features(VIRTIO_CAN_F_CAN_FD) {
+                warn!("Drop non-supported CAN FD frame");
+                return Err(Error::UnexpectedFdFlag);
+            }
+        } else if !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC) {
+            warn!("Drop non-supported CAN classic frame");
+            return Err(Error::UnexpectedClassicFlag);
+        }
+
+        // Add VIRTIO_CAN_FLAGS_EXTENDED in flag if the received frame
+        // had an extended CAN ID.
+        // Based on virtio-spec v1.4, the 3 MSB of can_id should be set to 0.
+        if (response.can_id.to_native() & CAN_EFF_FLAG) != 0 {
+            can_rx.flags = VIRTIO_CAN_FLAGS_EXTENDED.into();
+            can_rx.can_id = (response.can_id.to_native() & CAN_EFF_MASK).into();
+        } else {
+            can_rx.flags = 0.into();
+            can_rx.can_id = (response.can_id.to_native() & CAN_SFF_MASK).into();
+        }
+
+        // Remote transfer request is used only with classic CAN
+        if (response.can_id.to_native() & CAN_RTR_FLAG) != 0 {
+            if !self.check_features(VIRTIO_CAN_F_RTR_FRAMES)
+                || !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC)
+            {
+                warn!("Drop non-supported RTR frame");
+                return Err(Error::UnexpectedRtrFlag);
+            }
+            // If remote transfer request is enabled add the according flag
+            can_rx.flags = (can_rx.flags.to_native() | VIRTIO_CAN_FLAGS_RTR).into();
+        }
+
+        // Treat Vcan interface as CANFD if MTU is set to 64 bytes.
+        //
+        // Vcan can not be configured as CANFD interface, but it is
+        // possible to configure its MTU to 64 bytes. So if a messages
+        // bigger than 8 bytes is being received we consider it as
+        // CANFD message.
+        let can_name = self.controller.read().unwrap().can_name.clone();
+        if self.check_features(VIRTIO_CAN_F_CAN_FD)
+            && response.length.to_native() > 8
```

Comment:
I think you can minimize redundant conversions by storing the results in a variable.

Reply:
Thanks for the suggestion! I have it integrated